### DISCUSSION TITLE
Quorum queue setup on a datacenter with exactly two availability zones (racks)

### DISCUSSION THREAD

**Discussion by dev4342345235** [rabbitmq/rabbitmq-server#11877] [Questions] [answered]
Hi,

we would like to setup a RabbitMQ cluster using quorum queues in a single datacenter consisting of exactly two availability zones (=racks).

Unfortunately the RabbitMQ documentation is not clear to us, if this is a supported scenario. For example in [RackAwareness](https://www.rabbitmq.com/blog/2020/07/07/disaster-recovery-and-high-availability-101#rack-awareness) three Racks are mentioned, i.e. an odd number of racks. In [quorum-requirements](https://www.rabbitmq.com/docs/quorum-queues#quorum-requirements) it is mentioned that the cluster will not work, if the majority of nodes fails, which could happen with only two racks and an odd number of RabbitMQ nodes.

At the moment it seems, that it is not possible to setup a reliable quorum cluster spanning two racks.

Therefore please could you give us some insights what we could do?

Some ideas we were thinking about but not sure it they will work:

- In case of 3 quorum nodes add a fourth node to the rack hosting which hosts only one quorum node. In case of others rack fail add this "waiting node" to quorum group. But how to ensure, that quorum is always done on 2 nodes sitting in different racks?
- Use an even number of RabbitMQ node as quorum and therefore "force" RabbitMQ to sync to all nodes in all racks. Not sure if an even number is supported.

Edit: To clarify my question: My point is about failure handling of a whole Rack. All nodes on that Rack will immediately be unavailable. If that Rack would have hosted 2 of 3 RabbitMQ nodes, we only have one node left which is, as far as I understand, not a working condition for QQ.

The examples in the RabbitMQ blog (see link above) use an odd number of racks, this would avoid that issue, but at the moment we only have two racks.

Best
Chris

**Comment by michaelklishin** [comment]
@dev4342345235 it is perfectly possible to set up a reliable quorum queue that has three replicas, which means three cluster nodes. They can use two racks if that's your limit for any reason.

Two replica QQs are not supported. There is no such thing as a clear majority, which is very important for any practical Raft-based system, in a two node replica quorum queue for fairly obvious reasons.

**Reply by dev4342345235** [reply]
Sorry for unmarking this as answer as I still did not get the point of availability in case of Rack failure. Please see my additional comment.

Best 
Chris

**Reply by dev4342345235** [reply]
I have updated my original question to be more clear about my specific point.

Best
Chris

**Comment by michaelklishin** [comment]
Quorum queues will not be aware of racks. You need to make sure every rack has a RabbitMQ node deployed to it, and that there are three nodes total. Nothing beyond that. Exactly the same limitation and solution applies to streams.

**Comment by dev4342345235** [comment]
Thank you for you fast response!

Good to hear but I still don't get, how the cluster is able to survice if the rack goes down which hosts the majority of the nodes. Please take a look to the diagram from the [RabbitMQ website](https://www.rabbitmq.com/blog/2020/07/07/disaster-recovery-and-high-availability-101#rack-awareness):

With only two racks we would have a scenario as depicted in the first picture:

![image](https://github.com/user-attachments/assets/b331c6fb-6218-4eeb-bd97-3cb6546c79e3)

The documentation recommends to split the three nodes across three racks:

![image](https://github.com/user-attachments/assets/5c742638-dec1-46d7-9b44-95866ab60565)

So my question is, how we could achieve availability with two racks and a failure of the rack hosting the majority of the nodes? Do we need to go up to 5 nodes, put 3 nodes on Rack-A and 2 on Rack-B? But then we would still loose the majority of the nodes in case of Rack-A failure. From my understanding only 2 nodes are acceptable for failure if 5 are used in total.

Best
Chris

**Reply by michaelklishin** [reply, marked as answer]
With two racks, assuming an entire rack can fail, you cannot. You need to use three or assume that the risk of a rack failure is not important enough compared to host/node failure.

Two replica QQs and streams is an explicitly unsupported configuration (of course, you can extend a QQ or stream to just two replicas but it won't offer much in terms of availability).

**Reply by dev4342345235** [reply]
Ok, thank you for you super fast support here. :)

I was already afraid that with only two racks we will have an issue. As far as I understand you, the only way to go now is to convince the customer to add an additional rack, if they want to be prepared in case of full rack-faildown and still have a running cluster.

Best
Chris

**Reply by michaelklishin** [reply]
Like I said, full rack failure is a condition that many environments can consider to be too improbable to spend extra money on addressing. This is the majority of RabbitMQ clusters in the wild.

When that's not the case, then yes, three racks is the minimum since that's the minimal replication factor supported by modern replicated data types in RabbitMQ.

**Reply by kamilzzz** [reply]
What about having for example 5 node cluster with nodes spread between 3 racks.
As I can see there is no option to configure quorum queue members to be spread between racks/AZs/physical hosts.

In such case, how do I ensure specific quorum queue members does not end up being located on 2 racks which in the end would have the same end result - that rack going down resulting in queue unavailability (even though we had 3 racks, quorum members were located only on 2 of them, with one having 1 member, and another one having 2 members).

**Reply by dev4342345235** [reply]
By no means being a RabbitMQ expert by myself I would assume that in your case the queue is distributed to at least 3 of 5 nodes of the cluster and then one rack with 2 nodes could fail having one surving node on a different rack storing the data...  But definitly a RabbitMQ expert should comment on that

**Reply by michaelklishin** [reply]
@kamilzzz like I said above, quorum queues (or streams, or Khepri, or our Raft implementation) are not aware of racks so in your example, when it comes to rack-wide failure you will have the same [availability characteristics](https://www.rabbitmq.com/docs/quorum-queues#leader-election) as if you had only three nodes.

When it comes to other types of failures, e.g. disk, you will have a "true" five node cluster.

**Reply by kamilzzz** [reply]
> quorum queues (or streams, or Khepri, or our Raft implementation) are not aware of racks so in your example, when it comes to rack-wide failure you will have the same [availability characteristics](https://www.rabbitmq.com/docs/quorum-queues#leader-election) as if you had only three nodes

So if I understand correctly, with 3 "racks" (or cloud availability zones, or fault domains), out of the box availability characteristic of 5 nodes can actually be worse than with 3 node setup.
If I had only three nodes placed in different racks, quorum queue of size 3 would be guaranteed to span these racks.

With 5 nodes placed in 3 racks, there is a chance quorum queue replicas will span only 2 racks as there is no way to configure some topology aware hints regarding replica placement. And that leads to the issue were failure of a single rack can render queue broken.

I guess right now there is no better way than to manage replicas via CLI to ensure they are spanning 3 racks?

**Reply by michaelklishin** [reply]
If you assume that a failure will span entire availability zone or failure domain, yes, it can be. RabbitMQ quorum queues and streams are aware of how many replicas they have online, not in how many AZs or FDs they are deployed to.

**Reply by womblep** [reply]
@michaelklishin I have the exact same issue as the OP and no chance of adding a third rack/AZ. In the instance that I lose a complete rack/AZ I am going to lose either 1 node or 2 nodes. In both case RabbitMQ will be "down" i.e. not able to pass messages.
Assuming that I can spin up new nodes in the surviving rack/AZ, is the time that RabbitMQ is down limited to the time to spin up the new nodes and add them to the Quorum queues? And I wont lose the topology of the queues/exchanges/shovels/etc (as at least one cluster member is left)?

**Reply by michaelklishin** [reply]
@womblep if you lose one node it won't be "down" because 2 nodes out of 3 are a majority.

If you lose a majority you will have to [shrink all nodes to one replica](https://www.rabbitmq.com/docs/quorum-queues#replica-management), [remove all permanently lost nodes](https://www.rabbitmq.com/docs/clustering#removing-nodes) then add new ones and grow the number of replicas to three.

Simply adding nodes won't be enough. Not removing permanently gone nodes or [opting in](https://www.rabbitmq.com/docs/quorum-queues#replica-reconciliation) to have their replicas removed is also a mistake. Those automagical features, just like [forced removal of nodes not known to the peer discovery backend](https://www.rabbitmq.com/docs/cluster-formation#node-health-checks-and-cleanup) are sharp tools and should be considered carefully.

[Grow-then-Shrink Upgrades](https://www.rabbitmq.com/docs/upgrade#grow-then-shrink-upgrade) is a highly relevant documentation section.

**Reply by womblep** [reply]
@michaelklishin adding the new nodes with the previous node's disk would help mitigate that manual intervention, wouldn't it?
I.e Node A on Rack1, Nodes B & C on Rack2. Assuming disk is synced between racks, if Rack2 goes down then Nodes B & C could be spun up on Rack1 with the replica disk and should just continue as members of the cluster/quorum queues, correct?

**Reply by michaelklishin** [reply]
If the node names and identities are preserved, then those replicas will be considered existing Raft cluster members. Sometimes this assumption is reasonable and sometimes not.

Please read the guides linked to.
